#!/bin/bash

set -o errexit
set -o pipefail
set -o nounset


python << END
import sys
import time
import MySQLdb

suggest_unrecoverable_after = 30
start = time.time()

connection_string = "${DATABASE_URL}"
user = connection_string.split("//")[1].split(":")[0]
password = connection_string.split(":")[2].split("@")[0]
database = connection_string.split("/")[-1]
host = connection_string.split("@")[1].split(":")[0]
port = connection_string.split(":")[3].split("/")[0]

while True:
    try:
        db = MySQLdb._mysql.connect(
            host=host,
            user=user,
            password=password,
            database=database,
            port=int(port)
        )
        sys.stdout.write("MySQL is available.\n")
        break
    except MySQLdb._exceptions.OperationalError as error:
        sys.stderr.write("Waiting for MySQL to become available...\n")
        if time.time() - start > suggest_unrecoverable_after:
            sys.stderr.write("This is taking longer than expected. The following exception may be indicative of an unrecoverable error: '{}'\n".format(error))
            break

    time.sleep(1)
END


python << END
import sys
import time
import redis

suggest_unrecoverable_after = 30
start = time.time()

while True:
    try:
        r = redis.from_url("${REDIS_URL}")
        r.ping()
        sys.stdout.write("Redis is available.\n")
        break
    except redis.exceptions.RedisError as error:
        sys.stderr.write("Waiting for Redis to become available...\n")
        if time.time() - start > suggest_unrecoverable_after:
            sys.stderr.write("This is taking longer than expected. The following exception may be indicative of an unrecoverable error: '{}'\n".format(error))
            break

    time.sleep(1)
END

exec "$@"
